compiler/rustc_codegen_ssa/src/base.rs

```diff
@@ -348,12 +348,20 @@ pub fn maybe_create_entry_wrapper<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>>(
     cx: &'a Bx::CodegenCx,
 ) -> Option<Bx::Function> {
     let main_def_id = cx.tcx().entry_fn(LOCAL_CRATE).map(|(def_id, _)| def_id)?;
-    let instance = Instance::mono(cx.tcx(), main_def_id.to_def_id());
+    let main_is_local = main_def_id.is_local();
+    let instance = Instance::mono(cx.tcx(), main_def_id);
 
-    if !cx.codegen_unit().contains_item(&MonoItem::Fn(instance)) {
+    if main_is_local {
         // We want to create the wrapper in the same codegen unit as Rust's main
         // function.
-        return None;
+        if !cx.codegen_unit().contains_item(&MonoItem::Fn(instance)) {
+            return None;
+        }
+    } else {
+        // We want to create the wrapper only when the codegen unit is the primary one
+        if !cx.codegen_unit().primary() {
+            return None;
+        }
     }
```

Comment:
This logic will need to be adopted in cg_clif too. The required changes for this will likely conflict with changes on cg_clif's `master` branch, so I will open an issue on the cg_clif repo to implement it on the next rustup.